server/src/main/java/io/crate/planner/optimizer/rule/MoveConstantJoinConditionsBeneathJoin.java

```diff
@@ -36,39 +36,38 @@
 import io.crate.metadata.RelationName;
 import io.crate.metadata.TransactionContext;
 import io.crate.planner.consumer.RelationNameCollector;
-import io.crate.planner.operators.HashJoin;
+import io.crate.planner.operators.JoinPlan;
 import io.crate.planner.operators.LogicalPlan;
-import io.crate.planner.operators.NestedLoopJoin;
 import io.crate.planner.optimizer.Rule;
 import io.crate.planner.optimizer.costs.PlanStats;
 import io.crate.planner.optimizer.matcher.Captures;
 import io.crate.planner.optimizer.matcher.Pattern;
 import io.crate.sql.tree.JoinType;
 
-public class MoveConstantJoinConditionsBeneathNestedLoop implements Rule<NestedLoopJoin> {
+public class MoveConstantJoinConditionsBeneathJoin implements Rule<JoinPlan> {
 
-    private final Pattern<NestedLoopJoin> pattern;
+    private final Pattern<JoinPlan> pattern;
 
-    public MoveConstantJoinConditionsBeneathNestedLoop() {
-        this.pattern = typeOf(NestedLoopJoin.class)
-            .with(nl -> nl.joinType() == JoinType.INNER &&
-                        !nl.isJoinConditionOptimised() &&
-                        nl.joinCondition() != null);
+    public MoveConstantJoinConditionsBeneathJoin() {
+        this.pattern = typeOf(JoinPlan.class)
+            .with(join -> join.joinType() == JoinType.INNER &&
+                !join.isJoinConditionOptimized() &&
+                join.joinCondition() != null);
```

Comment:
```suggestion
             hasConstantJoinConditions(join.joinCondition()));
    }

    private static boolean hasConstantJoinConditions(Symbol joinCondition) {
        for (var condition : QuerySplitter.split(joinCondition).values()) {
            if (numberOfRelationsUsed(condition) <= 1) {
                return true;
            }
        }
        return false;
    }
```

You could put the check for the constant join conditions into the pattern and then the rule would not come up in the `explain verbose` by marking the join.